eth/downloader/queue.go

```diff
@@ -363,28 +383,28 @@ func (q *queue) WaitResults() []*fetchResult {
 		}
 		// Advance the expected block number of the first cache entry.
 		q.resultOffset += uint64(nproc)
+
+		// Recalculate the result item weights to prevent memory exhaustion
+		for _, result := range results {
+			var size common.StorageSize
+
+			blob, _ := rlp.EncodeToBytes(result.Receipts) // TODO(karalabe): Whoah, nice hack, super optimal
```

Comment:
You don't really need the encoded size here, any approximation of the size is enough. Maybe add a Size method to Receipt, Block, Transaction that returns the size of the struct + size of the content of slice fields.

Reply:
The transactions already had a Size method which worked based on RLP serialization. I've added a size method to the rest (which only approximates). PTAL